### ISSUE TITLE
TaskProcess never ends

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#5601]
TaskProcess is not stopping when DAG is over.
This has lead each schedd to saturate and no new TP can be started which is very bad of course. See picture.
![screenshot from 2017-11-06 11-38-32](https://user-images.githubusercontent.com/1837785/32444248-3a6b0542-c302-11e7-9578-eb4333a4bc8c.png)

**Comment by belforte** [comment]
AFAICT the problem is that the TP wrapper checks for DAG to be running using a query[1]  which includes TP iself in the list, so it lists two jobs, not one, confusing its own input processing [2]
resulting in [3]

[1]
https://github.com/dmwm/CRABServer/blob/master/scripts/task_process/task_proc_wrapper.sh#L41
[2]
https://github.com/dmwm/CRABServer/blob/master/scripts/task_process/task_proc_wrapper.sh#L25
[3]
```
belforte@vocms0122/task_process> tail -1 daemon.err.19923174.0/data/srv/glidecondor/condor_local/spool/3173/0/cluster19923173.proc0.subproc0/task_process/task_proc_wrapper.sh: line 26: (1509976480 - 1509078246 19923174 2 1509035621) < 24 * 3600 : missing `)' (error token is "19923174 2 1509035621) < 24 * 3600 ")
belforte@vocms0122/task_process> 
```

**Comment by belforte** [comment]
i.e. the problem is that this line
```
DAG_INFO=$(condor_q -constr 'CRAB_ReqName =?= "'$REQUEST_NAME'" && TaskType =!= "Job"' -af ClusterId JobStatus EnteredCurrentStatus)
TIME_OF_LAST_QUERY=$(date +"%s")
```
returns 
```
 echo $DAG_INFO 
19923173 4 1509078246 19923174 2 1509035621
```
therefore in line [1]
``` 
while read CLUSTER_ID DAG_STATUS ENTERED_CUR_STATUS; do
```
we end with `$ENTERED_CUR_STATUS` = `11509078246 9923174 2 1509035621` which creates the syntax error, but would not work in any case.

In other words $DAG_INFO as constructed is a single line with two jobs info. One way is of course to modifty it to query only the DAG (`TaskType=ROOT`) but I need to see how this works in the automatic splitting, multi-DAG case.
@matz-e any quick suggestion of advice ?

**Comment by matz-e** [comment]
@belforte Sorry, just read this after arriving in the US. `TaskType` for automatic splitting would be [defined here](https://github.com/dmwm/CRABServer/blob/181e86aecd112692e4107d97a21816a90cd13bf0/src/python/TaskWorker/Actions/PreDAG.py#L246) and can take values `TAIL` and `PROCESSING`.

**Comment by belforte** [comment]
so no |ROOT| anymore ? What about the probe DAG ?

**Comment by matz-e** [comment]
The probe DAG still uses the `ROOT` `TaskType`. I think there was some issue with re-using the `ROOT` type for the other DAGs, but I can't recall the details right now.

**Comment by matz-e** [comment]
OK, I think that the proper query part should be `stringListMember(TaskType, "ROOT PROCESSING TAIL", " ")` rather than `TaskType =!= "Job"` or `TaskType =?= "ROOT"`.

Since having the PROCESSING and TAIL DAG being direct dependents of the ROOT DAG led to problems with the resubmission of failed tail jobs, we severed the dependency chain, and ROOT, PROCESSING, and TAIL DAGs are all independent. This means:

* The ROOT DAG will not run over the whole duration of the task, but complete as soon as the processing jobs start (mostly within one hour, the DAG will be completed).
* The PROCESSING DAG will run longest, and will spawn the TAIL DAGs, and complete on the order of days. It may stop running and change status before all TAIL DAGs are done.
* The TAIL DAGs (a maximum of 3 currently) will run last, and shorter, complete on the order of a day.

**Comment by belforte** [comment]
thanks @matz-e I will change like that
And do those DAGs all run as jobs in the scheduler universe ? ( i.e. `JobUniverse==7`)

**Comment by matz-e** [comment]
Yes, they do. The task process actually runs in universe 12, I think. I'll have to verify the TAIL DAG behavior exactly still, but the code treats them the same way as PROCESSING DAGs.

**Comment by matz-e** [comment]
This is the output of #5622 applied:
```
+ condor_q -constr 'CRAB_ReqName =?= "180119_131422:matze_crab_0067_automatic_test_wrapper"'


-- Schedd: crab3-5@vocms059.cern.ch : <128.142.142.228:4080?... @ 01/22/18 11:25:39
 ID         OWNER            SUBMITTED     RUN_TIME ST PRI SIZE CMD
6397448.0   cms463          1/19 14:16   0+00:30:37 C  10   0.1 dag_bootstrap_startup.sh RunJobs.dag
6397621.0   cms463          1/19 14:46   0+01:05:26 C  10   0.3 condor_dagman -p 0 -f -l . -Lockfile RunJobs0.subdag.lock -Auto

2 jobs; 2 completed, 0 removed, 0 idle, 0 running, 0 held, 0 suspended
+ for i in '*67*'
+ set -x
+ condor_q -constr 'CRAB_ReqName =?= "180119_131653:matze_crab_0067_regular_test_wrapper"'


-- Schedd: crab3-5@vocms059.cern.ch : <128.142.142.228:4080?... @ 01/22/18 11:25:39
 ID         OWNER            SUBMITTED     RUN_TIME ST PRI SIZE CMD
6397455.0   cms463          1/19 14:19   0+20:05:45 C  10   0.1 dag_bootstrap_startup.sh RunJobs.dag

1 jobs; 1 completed, 0 removed, 0 idle, 0 running, 0 held, 0 suspended
+ for i in '*67*'
+ set -x
+ condor_q -constr 'CRAB_ReqName =?= "180119_131930:matze_crab_0067_automatic_test_totalunits"'


-- Schedd: crab3-5@vocms059.cern.ch : <128.142.142.228:4080?... @ 01/22/18 11:25:39
 ID         OWNER            SUBMITTED     RUN_TIME ST PRI SIZE CMD
6397610.0   cms463          1/19 14:21   0+00:25:36 C  10   0.1 dag_bootstrap_startup.sh RunJobs.dag
6397622.0   cms463          1/19 14:47   0+00:50:18 C  10   0.3 condor_dagman -p 0 -f -l . -Lockfile RunJobs0.subdag.lock -Auto

2 jobs; 2 completed, 0 removed, 0 idle, 0 running, 0 held, 0 suspended
+ for i in '*67*'
+ set -x
+ condor_q -constr 'CRAB_ReqName =?= "180119_135345:matze_crab_0067_automatic_test_wrapper2"'


-- Schedd: crab3-5@vocms059.cern.ch : <128.142.142.228:4080?... @ 01/22/18 11:25:39
 ID         OWNER            SUBMITTED     RUN_TIME ST PRI SIZE CMD
6397626.0   cms463          1/19 14:54   0+00:31:04 C  10   0.1 dag_bootstrap_startup.sh RunJobs.dag
6397657.0   cms463          1/19 15:25   0+08:50:11 H  10   0.3 condor_dagman -p 0 -f -l . -Lockfile RunJobs0.subdag.lock -Auto
6397992.0   cms463          1/19 21:34   0+02:49:40 C  10   0.3 condor_dagman -p 0 -f -l . -Lockfile RunJobs1.subdag.lock -Auto
6398001.0   cms463          1/20 00:16   0+01:48:17 C  10   0.3 condor_dagman -p 0 -f -l . -Lockfile RunJobs3.subdag.lock -Auto

4 jobs; 3 completed, 0 removed, 0 idle, 0 running, 1 held, 0 suspended
```
For automatic and regular splitting, the task process wrapper terminated (I submitted these before the weekend), with log outputs:
```
[2018-01-19 14:16] Starting a new task_process, creating task_process_running file
[2018-01-19 14:16] REQUEST_NAME: 180119_131422:matze_crab_0067_automatic_test_wrapper
...
[2018-01-20 15:52] Running an extra condor_q check before exitting
[2018-01-20 15:52] HTCondor query of DAG status done on 2018/01/20 15:52:53 CET
[2018-01-20 15:52] Dag status code for 6397448: 4 Entered current status date: 2018/01/19 14:46:57 CET
[2018-01-20 15:52] Dag status code for 6397621: 4 Entered current status date: 2018/01/19 15:52:23 CET
[2018-01-20 15:52] Dag(s) has (have) been in one of the final states for over 24 hours.
[2018-01-20 15:52] Caching the status one last time, removing the task_process/task_process_running file and exiting.
[2018-01-20 15:52] Running cache_status.py

[2018-01-19 14:19] Starting a new task_process, creating task_process_running file
[2018-01-19 14:19] REQUEST_NAME: 180119_131653:matze_crab_0067_regular_test_wrapper
...
[2018-01-21 10:26] Running an extra condor_q check before exitting
[2018-01-21 10:26] HTCondor query of DAG status done on 2018/01/21 10:26:32 CET
[2018-01-21 10:26] Dag status code for 6397455: 4 Entered current status date: 2018/01/20 10:24:59 CET
[2018-01-21 10:26] Dag(s) has (have) been in one of the final states for over 24 hours.
[2018-01-21 10:26] Caching the status one last time, removing the task_process/task_process_running file and exiting.
[2018-01-21 10:26] Running cache_status.py

[2018-01-19 14:21] Starting a new task_process, creating task_process_running file
[2018-01-19 14:21] REQUEST_NAME: 180119_131930:matze_crab_0067_automatic_test_totalunits
...
[2018-01-20 15:38] Running an extra condor_q check before exitting
[2018-01-20 15:38] HTCondor query of DAG status done on 2018/01/20 15:38:03 CET
[2018-01-20 15:38] Dag status code for 6397610: 4 Entered current status date: 2018/01/19 14:47:07 CET
[2018-01-20 15:38] Dag status code for 6397622: 4 Entered current status date: 2018/01/19 15:37:25 CET
[2018-01-20 15:38] Dag(s) has (have) been in one of the final states for over 24 hours.
[2018-01-20 15:38] Caching the status one last time, removing the task_process/task_process_running file and exiting.
[2018-01-20 15:38] Running cache_status.py

[2018-01-19 14:55] Starting a new task_process, creating task_process_running file
[2018-01-19 14:55] REQUEST_NAME: 180119_135345:matze_crab_0067_automatic_test_wrapper2
...
[2018-01-21 02:07] Running an extra condor_q check before exitting
[2018-01-21 02:07] HTCondor query of DAG status done on 2018/01/21 02:07:09 CET
[2018-01-21 02:07] Dag status code for 6397626: 4 Entered current status date: 2018/01/19 15:25:58 CET
[2018-01-21 02:07] Dag status code for 6397657: 5 Entered current status date: 2018/01/20 00:16:09 CET
[2018-01-21 02:07] Dag status code for 6397992: 4 Entered current status date: 2018/01/20 00:27:35 CET
[2018-01-21 02:07] Dag status code for 6398001: 4 Entered current status date: 2018/01/20 02:04:26 CET
[2018-01-21 02:07] Dag(s) has (have) been in one of the final states for over 24 hours.
[2018-01-21 02:07] Caching the status one last time, removing the task_process/task_process_running file and exiting.
[2018-01-21 02:07] Running cache_status.py
```
At least for my deployment, the task process wrapper quits on its own, and no additional cleanup is required.

**Comment by belforte** [comment]
I have found at least one reason why current task_process is sitll running for some tasks.
The problem is that the task_process job is restarted by condor, I can't say why, so that when the current script checks for "how long it has been running", it finds "less then a day" and keeps going... then job is restarted apparently every day..
```
001 (15293199.000.000) 12/17 10:33:59 Job executing on host: <188.185.81.0:4080?
...
001 (15293199.000.000) 01/19 23:20:30 Job executing on host: <188.185.81.0:4080?
...
001 (15293199.000.000) 01/20 03:31:25 Job executing on host: <188.185.81.0:4080?
...
001 (15293199.000.000) 01/21 20:53:24 Job executing on host: <188.185.81.0:4080?
...
001 (15293199.000.000) 01/22 02:14:58 Job executing on host: <188.185.81.0:4080?
...
001 (15293199.000.000) 01/22 02:46:47 Job executing on host: <188.185.81.0:4080?
...

```
That seems to call for some rewriting, but maybe you @matze have doene it already, I will check the code further for the old and new version.

**Comment by belforte** [comment]
Nope, code still relies on EnteredCurrentStatus so the logic is fooled by job restarts.
Of course task_process should not restart everyday :-( but I have no clue there

**Comment by belforte** [comment]
I spent a few hours with Todor learning about  bash variables and logic.
Eventually we came to the same conclusion as @matz-e that the orignal way to check if TP has been running for at least 24h after end of DAG was flawed a never going to work. Thanks @matz-e for having replaced a numerical expression containing `<` with a proper logical test using `-lt`.
For long term maintenance I would like to streamline the code even more to get rid of some mind twisting double negation, things like `NOT_DONE equal 0 means ...` and make the logic easier to follow. Anyhow... not urgent.

**Comment by matz-e** [comment]
This took me a while and quite some iterations, too. Regarding maintenance: I think that the code reads a bit better if you replace `NOT_DONE` by `ACTIVE`. Then we can exit when we have no `ACTIVE` DAGs…

**Comment by belforte** [comment]
I am giving it a try at a more readable version. Since I get confused so easily I am opening another issue. I like to keep this one closed.

**Comment by belforte** [comment]
I am surprised git did not close when I merged the PR. I will check and close by hand.

**Comment by belforte** [comment]
fixed via #5617 and #5622

**Comment by belforte** [comment]
(committitng this comment for history after a couple of weeks... sorry)
I spent a few hours with Todor learning about  bash variables and logic.
Eventually we came to the same conclusion as @matz-e that the orignal way to check if TP has been running for at least 24h after end of DAG was flawed a never going to work. Thanks @matz-e for having replaced a numerical expression containing `<` with a proper logical test using `-lt`.
For long term maintenance I would like to streamline the code even more to get rid of some mind twisting double negation, things like `NOT_DONE equal 0 means ...` and make the logic easier to follow. Anyhow... not urgent.
